### ISSUE TITLE
Cron runner module: sql state snapshots

### ISSUE THREAD

**Issue by cornelinux** [privacyidea/privacyidea#1028]
This is a cron runner module based on #992 

It can be used to save state snapshots to the monitoring database (#1027)

Such states could be something like #986 

* define the **state_name**
* define how the data is retrieved:
  * based on which database table
  * which kind of where statement should be used...
  * ``.count()`` is applied.

A runner that would ask "how many users with assigned tokens" might be difficult. This migh need an extra runner module.

**Comment by fredreichbier** [comment]
I guess the most difficult question is how we want to implement data retrieval, i.e. how the user should specify the quantity to monitor. Some possibilities:

1) Use a list of predefined queries (e.g. "how many tokens", "how many users with assigned tokens"). Easiest to implement, but pretty boring. :-)

2) User can specify a table and a simple WHERE clause, e.g. "Token" and "active == 1" or "failcount > 5", from which privacyIDEA builds a query and applies ``.count()``. This would already allow some interesting queries but everything involving users would be out of reach because users aren't stored in a table. We could, however, invent some "virtual" table that mimics an SQL table. So the user could choose the virtual "User" table and specify the WHERE clause "resolver = foobar AND assigned = 0", which the snapshot module would internally convert not to a SQL SELECT clause, but to a resolver ``getUserList()`` call with some subsequent filtering. This would allow many interesting queries, but could get a bit clunky.

**Comment by cornelinux** [comment]
In the token table there is the userid and the resolver. So it might be intersting to be also able to use some ``unique`` SQL statements.

Also, there is information, that is only available in the ``tokeninfo`` table. This would only be accessable via a ``join`` of the token and the tokeninfo table.

Maybe we could define the enhanced queries based on the classes, instead of the SQL tables.

**Comment by fredreichbier** [comment]
Yes, sounds good! I just wrote a simple SQL-like query parser based on pyparsing today which we could use to formulate queries based on classes. It is able to parse complex filters like
```sql
column == "some value" OR (some_other_column != easyvalue AND a == c)
```

**Comment by cornelinux** [comment]
We remove this from milestone 2.23. We start with a SimpleStats Cronrunner Module.

**Comment by cornelinux** [comment]
I think we should actually provide a script cron runner module and an example script, how we can easily fetch information with python/sqlalchemy. If we then see, that certain constellations are used over and over again, we can try to do it in the UI directly.
